File: cli/src/commands/status.ts
/**
 * mrsf status — Show anchor health for all comments.
 */

import type { Command } from "commander";
import chalk from "chalk";
import path from "node:path";
import { sidecarToDocument } from "../lib/discovery.js";
import { parseSidecar, readDocumentLines } from "../lib/parser.js";
import { resolveSidecarPaths } from "../lib/resolve-files.js";
import { findRepoRoot, getCurrentCommit, isStale as gitIsStale } from "../lib/git.js";
import { exactMatch } from "../lib/fuzzy.js";
import type { AnchorHealth, StatusResult } from "../lib/types.js";

export function registerStatus(program: Command): void {
  program
    .command("status [files...]")
    .description("Show anchor health for comments")
    .option("--json", "Output as JSON")
    .action(async (files: string[], opts: { json?: boolean }) => {
      const parentOpts = program.opts();
      const cwd = parentOpts.cwd ?? process.cwd();
      const quiet = parentOpts.quiet ?? false;
      const verbose = parentOpts.verbose ?? false;

      const sidecarPaths = await resolveSidecarPaths(files, cwd);

      if (sidecarPaths.length === 0) {
        if (!quiet) console.log("No sidecar files found.");
        return;
      }

      const repoRoot = await findRepoRoot(cwd);
      const allResults: StatusResult[] = [];

      for (const sp of sidecarPaths) {
        try {
          const doc = await parseSidecar(sp);
          const docPath = sidecarToDocument(sp);
          let lines: string[];
          try {
            lines = await readDocumentLines(docPath);
          } catch {
            // Document missing
            for (const c of doc.comments) {
              allResults.push({
                commentId: c.id,
                health: "unknown",
                reason: "Document file not found.",
              });
            }
            continue;
          }

          for (const c of doc.comments) {
            const result = await assessHealth(c, lines, repoRoot);
            allResults.push(result);
          }
        } catch (err: unknown) {
          const msg = err instanceof Error ? err.message : String(err);
          console.error(chalk.red(`✗ ${sp}: ${msg}`));
        }
      }

      if (opts.json) {
        console.log(JSON.stringify(allResults, null, 2));
        return;
      }

      if (verbose) {
        for (const sp of sidecarPaths) {
          console.log(chalk.dim(`  sidecar: ${sp}`));
        }
      }

      for (const r of allResults) {
        const icon = healthIcon(r.health);
        console.log(`${icon} ${r.commentId}: ${r.reason}`);
      }

      const fresh = allResults.filter((r) => r.health === "fresh").length;
      const stale = allResults.filter((r) => r.health === "stale").length;
      const orphaned = allResults.filter((r) => r.health === "orphaned").length;
      const unknown = allResults.filter((r) => r.health === "unknown").length;

      if (!quiet) {
        console.log(
          `\n${allResults.length} comment(s): ${fresh} fresh, ${stale} stale, ${orphaned} orphaned, ${unknown} unknown`,
        );
      }
    });
}

async function assessHealth(
  comment: { id: string; commit?: string; selected_text?: string; x_reanchor_status?: unknown },
  lines: string[],
  repoRoot: string | null,
): Promise<StatusResult> {
  // Already marked orphaned
  if (comment.x_reanchor_status === "orphaned") {
    return {
      commentId: comment.id,
      health: "orphaned",
      reason: "Marked orphaned by previous re-anchor.",
    };
  }

  // Check text match
  if (comment.selected_text) {
    const matches = exactMatch(lines, comment.selected_text);
    if (matches.length === 0) {
      // No exact match — might be stale or orphaned
      if (comment.commit && repoRoot) {
        const stale = await gitIsStale(comment.commit, repoRoot);
        if (stale) {
          return {
            commentId: comment.id,
            health: "stale",
            reason: "Commit differs from HEAD and text not found. Run reanchor.",
          };
        }
      }
      return {
        commentId: comment.id,
        health: "orphaned",
        reason: "Selected text not found in current document.",
      };
    }

    // Text found — check commit freshness
    if (comment.commit && repoRoot) {
      const stale = await gitIsStale(comment.commit, repoRoot);
      if (stale) {
        return {
          commentId: comment.id,
          health: "stale",
          reason: "Text still matches but commit is behind HEAD.",
        };
      }
    }

    return {
      commentId: comment.id,
      health: "fresh",
      reason: "Text matches in current document.",
    };
  }

  // No selected_text
  if (!comment.commit) {
    return {
      commentId: comment.id,
      health: "unknown",
      reason: "No selected_text or commit to assess.",
    };
  }

  if (repoRoot) {
    const stale = await gitIsStale(comment.commit, repoRoot);
    return {
      commentId: comment.id,
      health: stale ? "stale" : "fresh",
      reason: stale
        ? "Commit is behind HEAD (no text to verify)."
        : "Commit matches HEAD.",
    };
  }

  return {
    commentId: comment.id,
    health: "unknown",
    reason: "Git not available for commit check.",
  };
}

function healthIcon(health: AnchorHealth): string {
  switch (health) {
    case "fresh":
      return chalk.green("●");
    case "stale":
      return chalk.yellow("◐");
    case "orphaned":
      return chalk.red("✗");
    case "unknown":
      return chalk.dim("?");
  }
}

File: python/src/mrsf/cli/status.py
"""mrsf status [files...] — check anchor health of comments."""

from __future__ import annotations

import json
import sys

import click

from ..git import find_repo_root, get_current_commit, is_git_available
from ..parser import parse_sidecar, read_document_lines
from ..resolve_files import resolve_sidecar_paths
from ..types import Comment
from .main import MrsfContext, pass_ctx


def _assess_health(
    comment: Comment,
    doc_lines: list[str] | None,
    head_commit: str | None,
) -> tuple[str, str]:
    """Return (status, reason) for a comment's anchor health."""
    has_text = comment.selected_text is not None and len(comment.selected_text) > 0
    has_commit = comment.commit is not None and len(comment.commit) > 0
    is_orphaned = comment.extra.get("x_reanchor_status") == "orphaned"

    if is_orphaned:
        return "orphaned", "marked orphaned by reanchor"

    if not has_text and not has_commit:
        return "unknown", "no anchor data"

    if has_text and doc_lines is not None and comment.selected_text is not None:
        text_found = comment.selected_text in "\n".join(doc_lines)
        if text_found:
            if has_commit and head_commit and comment.commit == head_commit:
                return "fresh", "text matches, commit is HEAD"
            elif has_commit:
                return "stale", "text matches, commit is not HEAD"
            else:
                return "fresh", "text matches"
        else:
            if has_commit and head_commit and comment.commit == head_commit:
                return "stale", "text not found, commit is HEAD"
            else:
                return "stale", "text not found in document"

    if has_commit and head_commit:
        if comment.commit == head_commit:
            return "fresh", "commit is HEAD"
        else:
            return "stale", "commit is not HEAD"

    return "unknown", "insufficient anchor data"


@click.command("status")
@click.argument("files", nargs=-1, type=click.Path())
@click.option("--json", "as_json", is_flag=True, help="Output as JSON.")
@pass_ctx
def status_cmd(ctx: MrsfContext, files: tuple[str, ...], as_json: bool) -> None:
    """Check anchor health of sidecar comments."""
    sidecars = resolve_sidecar_paths(list(files), cwd=ctx.cwd)
    if not sidecars:
        click.echo("No sidecar files found.", err=True)
        sys.exit(1)

    head_commit = None
    if is_git_available():
        try:
            repo_root = find_repo_root(ctx.cwd)
            if repo_root:
                head_commit = get_current_commit(repo_root)
        except Exception:
            pass

    all_results: list[dict[str, str]] = []
    for sc in sidecars:
        try:
            doc = parse_sidecar(sc)
        except Exception as e:
            if not ctx.quiet:
                click.echo(click.style(f"Error parsing {sc}: {e}", fg="red"), err=True)
            continue

        # Read document lines for text matching
        doc_lines = None
        import os

        doc_dir = os.path.dirname(sc)
        doc_path = os.path.join(doc_dir, doc.document)
        if os.path.isfile(doc_path):
            try:
                doc_lines = read_document_lines(doc_path)
            except Exception:
                pass

        counts = {"fresh": 0, "stale": 0, "orphaned": 0, "unknown": 0}
        for c in doc.comments:
            status, reason = _assess_health(c, doc_lines, head_commit)
            counts[status] = counts.get(status, 0) + 1

            if as_json:
                all_results.append(
                    {
                        "sidecar": sc,
                        "comment_id": c.id,
                        "status": status,
                        "reason": reason,
                    }
                )
            elif not ctx.quiet:
                icons = {"fresh": "●", "stale": "◐", "orphaned": "✗", "unknown": "?"}
                colors = {
                    "fresh": "green",
                    "stale": "yellow",
                    "orphaned": "red",
                    "unknown": "white",
                }
                icon = icons.get(status, "?")
                color = colors.get(status, "white")
                click.echo(click.style(f"  {icon} {c.id}: {reason}", fg=color))

        if not as_json and not ctx.quiet:
            total = len(doc.comments)
            click.echo(
                f"  {sc}: {total} comment(s): "
                f"{counts['fresh']} fresh, {counts['stale']} stale, "
                f"{counts['orphaned']} orphaned, {counts['unknown']} unknown"
            )

    if as_json:
        click.echo(json.dumps(all_results, indent=2))
